puzzlejs/puzzle.js

```diff
@@ -2037,22 +2068,43 @@ function PuzzleGrid(puzzleEntry, options, container, puzzleId) {
                 var acrossClue = (c == 0 || textLines[r][c-1] == '@' || (edgeCode & 4)) && c < textLines[r].length - 1 && textLines[r][c+1] != '@' && !(edgeCode & 8); // block/edge left, letter right
                 var downClue = (r == 0 || textLines[r-1][c] == '@' || (edgeCode & 1)) && r < textLines.length - 1 && textLines[r+1][c] != '@' && !(edgeCode & 2); // block/edge above, letter below
                 
-                if (acrossClue || downClue || this.options["data-clue-locations"] == "all") {
+                if (acrossClue || downClue || this.options["data-clue-locations"] == "all" || this.options["data-clue-locations"] == "crostic") {
                     var clueIndicator = (!clueIndicators) ? ++clueNum : clueIndicators[clueNum++];
                     const isClueEmpty = String(clueIndicator).trim() === "";
 
-                    if (!isClueEmpty && this.options["data-clue-locations"] == "crossword") {
-                        if (acrossClue) { td.setAttribute("data-across-cluenumber", clueIndicator); }
-                        if (downClue) { td.setAttribute("data-down-cluenumber", clueIndicator); }
-                        if (acrossClue && acrossClues[acrossClueIndex]) {
-                          acrossClues[acrossClueIndex].setAttribute("data-across-cluenumber", clueIndicator);
-                          acrossClues[acrossClueIndex].setAttribute("value", clueIndicator);
-                          acrossClueIndex++;
+                    if (!isClueEmpty) {
+                        if (this.options["data-clue-locations"] == "crossword") {
+                            if (acrossClue) { td.setAttribute("data-across-cluenumber", clueIndicator); }
+                            if (downClue) { td.setAttribute("data-down-cluenumber", clueIndicator); }
+                            if (acrossClue && acrossClues[acrossClueIndex]) {
+                              acrossClues[acrossClueIndex].setAttribute("data-across-cluenumber", clueIndicator);
+                              acrossClues[acrossClueIndex].setAttribute("value", clueIndicator);
+                              acrossClueIndex++;
+                            }
+                            if (downClue && downClues[downClueIndex]) {
+                              downClues[downClueIndex].setAttribute("data-down-cluenumber", clueIndicator);
+                              downClues[downClueIndex].setAttribute("value", clueIndicator);
+                              downClueIndex++;
+                            }
                         }
-                        if (downClue && downClues[downClueIndex]) {
-                          downClues[downClueIndex].setAttribute("data-down-cluenumber", clueIndicator);
-                          downClues[downClueIndex].setAttribute("value", clueIndicator);
-                          downClueIndex++;
+                        if (this.options["data-clue-locations"] == "crostic") {
```

Comment:
There are some things that I like very much here:

- The description of linking IDs with multiple parts and knowing what to do with those parts
- The ability to simply move overlay components (clues, extracts, future things) around with no added CSS necessary

That said, I am not keen on merging the concepts of links (what I have called "extracts", but I think "links" is very very evocative and I am regretting my name choice!) and clues. I have already used both in the same cell in several puzzles (e.g. a crossword or KenKen where highlighted cells extract an answer) and feel that this would be limiting. 

I'd like to try this a slightly different way than what you've done here, which I believe preserves the spirit of what you're after:

- the ability to specify something like `data-extracts="C|1 A|2 C|3 C|4 C|5 B|6 C|7 B|8 B|9 B|10 A|11 C|12 A|13"`, where the linking codes are 1/2/3 etc but the displayed text is C1/A2/C3 etc. This is similar to what you're doing in data-clue-indicators but keeps the functionality in data-extracts and matches the layout that I see in crostic puzzles elsewhere.
- properties like `data-extract-position="top-right"`, which would add a `pos-top-right` class to all extracts (and have good defaults in the CSS). The CSS properties would also directly transfer to a `data-clue-position` property as well.
- a similar `avoid-top-right`, etc. class for cell contents, perhaps automatically set when there are clues or extracts.
- `data-modes` like `crostic-body` and `crostic-clue` that set sensible defaults from the above.

So it would look something like:

```
<div class="puzzle-entry" data-mode="crostic">
    <div style="min-width: 400px; background-color: white; padding: 10px;">
        <div class="puzzle-grid" data-mode="crostic-body" data-text="#####@#|#######"
            data-extracts="C|1 A|2 C|3 C|4 C|5 B|6 C|7 B|8 B|9 B|10 A|11 C|12 A|13"></div>
        <br>
        <div style="display: grid; grid-template-columns: 160px auto; row-gap: 20px; align-items: center;">
            <div>A. Not new</div>
            <div class="puzzle-grid" data-mode="crostic-clue" data-text="###" data-extracts="11 2 13"></div>
            <div>B. Left, on a map</div>
            <div class="puzzle-grid" data-mode="crostic-clue" data-text="####" data-extracts="10 6 8 9"></div>
            <div>C. Lights, Camera, ____!</div>
            <div class="puzzle-grid" data-mode="crostic-clue" data-text="######" data-extracts="7 1 5 3 12 4"></div>
        </div>
    </div>
</div>
```

I recognize that this is a lot and I am not asking you to do all of it, but I'd like to know if it covers all of the issues you're trying to address in a similarly straightforward way.

Does that get everything?

Reply:
I had actually added links as a new concept, specifically so that I wasn't using extracts, leaving that still available. (I tested by created a crostic that also had extracts.) I think your version won't support having an extract of just some cells from the crostic. Am I missing something?

Reply:
The issue is that it appears that you are using `data-clue-indicators` to hold things that are meant to be links, whereas I intend `data-extracts` to hold the links. [I agree that my property naming isn't the absolute best for carrying out this discussion, but semantically this is what I am after.] That said, perhaps *I* am missing something.

Reply:
I really like the concept of being able to specify overlay location (data-clue-position, data-extract-position, etc) without needing to use CSS!

Reply:
Having crostic-body and crostic-clue as separate types seems nice. Great that the different look for each of them could be automatic from those names instead of needing to set the layout details on each puzzle-grid div.

Reply:
Thanks - and I am trying to specify it in a puzzle-type-agnostic way, so that puzzle types we've never even thought of can benefit.

Reply:
> The issue is that it appears that you are using `data-clue-indicators` to hold things that are meant to be links, whereas I intend `data-extracts` to hold the links. [I agree that my property naming isn't the absolute best for carrying out this discussion, but semantically this is what I am after.] That said, perhaps _I_ am missing something.

Maybe using 'data-clue-indicators' isn't the best solution, but I was trying to be as minimally disruptive as possible. My goal was to have a way to indicate in two different puzzle-grid DIVs that a given '#' represented the same answer letter. I originally used data-extracts (from your crostic example), but I didn't like that it prevented using that feature for extracting a few cells from the grid for a final answer. (Not that I need that now, just thinking of future use.)

So, I chose to consider the data-clues as the identifier that should match across grids. I added the new data-link-id to accomplish that, so the extract feature is still possible for a crostic grid. (data-link-id works like data-extract-id to keep the cells in sync). 

Ultimately, if you don't like the idea of expanding clues to something that can be linked, it sounds like adding a new link-like parameter would be a solution.

One other comment related to crostics -- my personal preference is to leave the letter off for each of the individual clues, which is what necessitates having something that understands the concept of number+letter in the main grid with the letter essentially just being a decoration; the actual linking is based purely on the numbers (in both grids). That really feels to me like a clue-indicator, which is why I chose that implementation. That said, I'm fine with however you want to implement it!

Reply:
Oh! Here is what I am getting wrong about this. I grew up solving puzzles such as crostics in GAMES Magazine, which used a different format than seems to be in use today!

In GAMES Magazine, the number and letter were fused together into a single item (though I checked my back issues and see that I got the order backwards; it's 1A, not A1). I see that in the NYT, New Yorker, and elsewhere, the format is consistently different from this, and that format is what I assume you are trying to replicate. I haven't done a lot of crostics in the last 20+ years and have not noticed the change in standard (or maybe GAMES was always nonstandard and I just never knew). I now 100% understand why you are saying that the number looks like a clue indicator (though I still don't want to think of it as a clue indicator). 

Here is a new version that I think will make everything work and also be super flexible; I've only changed the `crostic-body` element. I've explicitly written the `data-extract-position` attribute below to help explain, though this would really be folded into `data-mode="crostic"` and would not generally need to be specified.

```
<div class="puzzle-entry" data-mode="crostic">
    <div style="min-width: 400px; background-color: white; padding: 10px;">
        <div class="puzzle-grid" data-mode="crostic-body" data-text="#####@#|#######"
            data-extract-position="top-left|top-right"
            data-extracts="1|C 2|A 3|C 4|C 5|C 6|B 7|C 8|B 9|B 10|B 11|A 12|C 13|A"></div>
        <br>
        <div style="display: grid; grid-template-columns: 160px auto; row-gap: 20px; align-items: center;">
            <div>A. Not new</div>
            <div class="puzzle-grid" data-mode="crostic-clue" data-text="###" data-extracts="11 2 13"></div>
            <div>B. Left, on a map</div>
            <div class="puzzle-grid" data-mode="crostic-clue" data-text="####" data-extracts="10 6 8 9"></div>
            <div>C. Lights, Camera, ____!</div>
            <div class="puzzle-grid" data-mode="crostic-clue" data-text="######" data-extracts="7 1 5 3 12 4"></div>
        </div>
    </div>
</div>
```

This would split the extract code into two components and separately position them; it would put the numbers in the top left of each cell and the letters in the top right of each cell, but the placement could be changed for whatever purpose the author deemed fit; any number of parts in any location deemed desirable. A screenreader would read out all the parts concatenated together. The first term of the code (in this case, the number) would be the true extract code for linking purposes, and that choice could be straightforwardly extended further in the future if desired (I have an idea here but don't want to confuse things at present).

The default value of `data-extract-position` would be `bottom-right` to match current behavior in non-crostic scenarios. The `crostic-body` and `crostic-clue` modes would both overwrite that default.

Reply:
That's getting closer to what I envision. I like the "|" notation to indicate multiple parts for the cell label and having the parallel 'position' param. (I'd been using '_' but '|' seems better.)
The issue is that it's still 'stealing' the extract feature, which means that we can't extract letters from the grid for a final answer.
I'm going to add a comment at the end of this page, as part of the general conversation (instead of on a specific piece of code) to start talking about the individual features and implementation ideas.

Reply:
I think the multiple-levels-of-linking issue is going to be solvable with creative marking of extraction labels, especially if one of the values in `data-extract-position` is `none`. Then I can use a single underlying code across all three locations in your scenario, and display two different types of linking for two slightly different purposes, all in the same single property. I can work out a full example of this if it isn't clear.

Whether the base property name should be "extract" or "link" is a separate matter, of course. I'm not ready for the breaking change at present, but if I receive more feedback like yours I could get there.

Reply:
Adding another extract value and using none is a great solution. (No need to change the name to link.)
Then my only other issue is that I don't want the yellow coloring for all the cells, but I can override that with CSS on my puzzle's page.